### ISSUE TITLE
python driver doesn't connect to all available IPs for a given DNS address 

### ISSUE THREAD

**Issue by avnerbarr** [grpc/grpc#20146]
### What version of gRPC and what language are you using?

python
grpcio==1.18.0
grpcio-tools==1.18.0

### What operating system (Linux, Windows,...) and version?
centos 6

### What runtime / compiler are you using (e.g. python version or version of gcc)
don't know

### What did you do?
I have a grpc service running on ECS
the service has an autoscale policy based on CPU usage and will from time to time increase/decrease number of tasks running.

the tasks are reporting request metrics and I see that only some of the instance are receiving requests.

From within a client machine I ran some checks:

service is on port :9999

`netstat -anp | grep 9999`

I see only a single IP:

```
netstat -anp | grep 9999
tcp6       0      0 172.30.7.243:49890      172.30.3.138:9999       ESTABLISHED 3756/python
tcp6       0      0 172.30.7.243:49902      172.30.3.138:9999       ESTABLISHED 4123/python
tcp6       0      0 172.30.7.243:49886      172.30.3.138:9999       ESTABLISHED 3922/python
tcp6       0      0 172.30.7.243:49922      172.30.3.138:9999       ESTABLISHED 4196/python
tcp6       0      0 172.30.7.243:49914      172.30.3.138:9999       ESTABLISHED 3576/python
tcp6       0      0 172.30.7.243:49910      172.30.3.138:9999       ESTABLISHED 4255/python
tcp6       0      0 172.30.7.243:49898      172.30.3.138:9999       ESTABLISHED 4026/python
```

from the same machine i ran `dig` on the dns of the service:

```
> dig my-service.local

...

;; QUESTION SECTION:
;my-service.local.	IN	A

;; ANSWER SECTION:
my-service.local. 60	IN	A	172.30.7.31
my-service.local. 60	IN	A	172.30.3.138

;; Query time: 1 msec
;; SERVER: 172.30.0.2#53(172.30.0.2)
;; WHEN: Sun Sep 01 06:50:19 UTC 2019
;; MSG SIZE  rcvd: 74
```

In other words the driver is only connecting to a single IP while there are 2 that should be connected.

If I go ahead and restart the client (kill the PID and run again), I do see that the 2 IPS are then connected to as expected

My connection is configured like so:

```python
        opts = [("grpc.lb_policy_name", "round_robin"), ("grpc.dns_enable_srv_queries", 1)]
        self.channel = grpc.insecure_channel(service_endpoint, opts)
        self.stub = MyServiceStub(self.channel)
```

### What did you expect to see?
All IPS should be receiving traffic

### What did you see instead?
Only a single IP receives traffic

**Comment by gnossen** [comment]
Can you provide more information about the lifetime of your channel? Do you reuse the same channel to send multiple RPCs? Or do you only send a single RPC on each channel before tearing it down?

**Comment by avnerbarr** [comment]
the channel is long living through out the lifetime of the application. the channel is never explicitly closed/torn down

**Comment by gnossen** [comment]
CC @apolcyn

**Comment by gnossen** [comment]
You mentioned `SRV` records, but your `dig` output shows `A` records. `SRV` records are [only intended for use with gRPCLB](https://github.com/grpc/proposal/blob/master/A5-grpclb-in-dns.md).  Are you actually trying to use gRPCLB?

**Comment by apolcyn** [comment]
From:

```
I have a grpc service running on ECS
the service has an autoscale policy based on CPU usage and will from time to time increase/decrease number of tasks running.
```

and

```
If I go ahead and restart the client (kill the PID and run again), I do see that the 2 IPS are then connected to as expected
```

One guess is, was there only a single IP available when you first started the "driver" ?

Note that if you have a grpc client channel that has connections which are "healthy", it won't retry DNS resolution unless provoked to do so (basically the client will only try to refresh DNS resolution  one of your connections goes bad). So in steady state your client won't "learn" about the new IP address as soon as it's available in DNS.

If that's not the case, suggest sharing a log from running your driver under environment variables `GRPC_VERBOSITY=DEBUG`, and `GRPC_TRACE=api,cares_resolver,cares_address_sorting,client_channel_routing`

**Comment by avnerbarr** [comment]
@gnossen 
I didn't mention using srv records explicitly. I assume it was implied from the the connection configuration. But in general "dns_enable_srv_queries" shouldn't deduce "dns_USE_ONLY_srv_queries" or some semantic like that. In general all acts normally until a scale event occurs.

(I think the documentation is completely lacking in this regards along with best practice and I arrived at that configuration from a seperate service which does have SRV records enabled (along with A records) and digging around in the C headers)

I can remove "dns_enable_srv_queries" if necessary - or alternatively I can add SRV records for this DNS name as well.

But regarding the SRV records, other services which have them enabled suffer from exactly the same problem. The client will not pick up new instances once the services are scaled. Easily reproduced using the same 'netstat + dig' from above.

**Comment by josephharrington** [comment]
Just wanted to +1 this issue to keep it alive. We use DNS-based load balancing so the list of resolved IPs can change frequently.
> Note that if you have a grpc client channel that has connections which are "healthy", it won't retry DNS resolution unless provoked to do so (basically the client will only try to refresh DNS resolution one of your connections goes bad).

"...unless provoked to do so" -- Is there a way to do this outside of recreating the channel or failing connection to all resolved IPs? Currently we have channel wrapper embedded into our Python clients that creates a new channel every 30 seconds. New calls will use the most recent channel and we keep the old channels around for a minute or two to allow inflight calls complete. Is there a better way to trigger DNS re-resolution? Ideally all we'd need to do is set appropriate TTLs on DNS records (and I believe we are already doing that). But alternatively it would be nice if there was a hook to trigger DNS re-resolution, so we wouldn't need the extra complexity of channel rotation.

**Comment by markdroth** [comment]
I think this is a duplicate of #12295.
